### ISSUE TITLE
cannot mmap an empty file error with parallel=True

### ISSUE THREAD

**Issue by falkben** [seung-lab/cloud-volume#164]
I receive the following error under certain circumstances when I add data to a CloudVolume instance when parallel is set to `True`.  When parallel is set to `False`, I do not get an error.  It seems to happen when the data is not "chunk" aligned?

If this isn't clear what I'm doing from below, I'd be happy to give more information.

```sh
Traceback (most recent call last):
  File "create_precomputed.py", line 75, in <module>
    vol[0:x, 0:y, 16:20] = rawdata[:,:,:]
  File "/home/ubuntu/env/lib/python3.6/site-packages/cloudvolume/cloudvolume.py", line 1059, in __setitem__
    txrx.upload_image(self, img, bbox.minpt, parallel=self.parallel)
  File "/home/ubuntu/env/lib/python3.6/site-packages/cloudvolume/txrx.py", line 298, in upload_image
    manual_shared_memory_order=manual_shared_memory_order)
  File "/home/ubuntu/env/lib/python3.6/site-packages/cloudvolume/txrx.py", line 336, in upload_aligned
    location=shared_memory_id, order=manual_shared_memory_order, lock=fs_lock)
  File "/home/ubuntu/env/lib/python3.6/site-packages/cloudvolume/sharedmemory.py", line 73, in ndarray
    return ndarray_shm(shape, dtype, location, readonly, order, **kwargs)
  File "/home/ubuntu/env/lib/python3.6/site-packages/cloudvolume/sharedmemory.py", line 173, in ndarray_shm
    array_like = mmap.mmap(shared.fd, shared.size)
ValueError: cannot mmap an empty file
```

**Comment by william-silversmith** [comment]
Hi Ben,

Could you share the following information: OS, CloudVolume version, whether you're running in a docker container, the size, offset, and chunksize of your dataset, and construction and CloudVolume call? 

One thing you can try is checking (on Ubuntu) whether there are lingering cloudvolume shared memory files in `/dev/shm/` and deleting them. They could be sucking up the space you need for new allocations.

Thanks!
Will

**Comment by falkben** [comment]
OS: Ubuntu 18.04.1
CloudVolume ver:  0.40.1
Not in docker
Extents of data are 2160 x 2560 x 1454
No offset
Chunksize: 128x128x16
CloudVolume path is to s3

```python
info = CloudVolume.create_new_info(
    num_channels=1,
    layer_type='image',
    data_type='uint16',  # Channel images might be 'uint8'
    encoding='raw',  # raw, jpeg, compressed_segmentation, fpzip, kempressed
    resolution=[x_size, y_size, z_size],  # Voxel scaling, units are in nanometers
    voxel_offset=[0, 0, 0],  # x,y,z offset in voxels from the origin
    # Pick a convenient size for your underlying chunk representation
    # Powers of two are recommended, doesn't need to cover image exactly
    chunk_size=[128, 128, 16],  # units are voxels
    # e.g. a cubic millimeter dataset
    volume_size=[x_size*x, y_size*y, z_size*z]
)
```

```python
vol = CloudVolume(cloud_volume_path, info=info, non_aligned_writes=True, fill_missing=True,
                  progress=False, parallel=True)
```

I'm iterating over z slices, and modulo 16 (my thinking was that this would be preferred as it matches the chunk size) commit them like so:

```python
vol[0:x, 0:y, tif_idx-15:tif_idx+1] = rawdata[:, :, :]
```

Since this volume isn't aligned to a chunk of z, I load the final slices and try to do a final commit, and that's when I finally get the error, *if parallel is set to True*

**Comment by william-silversmith** [comment]
I think I understand what's going on. `parallel` and `non_aligned_writes` have an interesting interaction. Currently, `parallel` only is applied to the aligned core of a non-aligned cutout. The non-aligned shell is handled by single-process code. I think what is happening is that the size of your cutout includes no aligned core, and errors when it allocates a buffer of zero size. This is a case that CloudVolume should handle gracefully. I'll submit a fix for it soon.

However, for your use case, the parallel flag won't do anything for you, so it might be best to just turn it off. It's possible to do more to parallelize the processing of the non-aligned shell and perhaps I should revisit it.

I hope that helps!
Will

**Comment by falkben** [comment]
That's interesting you say that -- I definitely notice a difference when using parallel vs. not.

![image](https://user-images.githubusercontent.com/653031/49969612-f45dd480-fef6-11e8-8fc1-fcecdb70cd85.png)

**Comment by william-silversmith** [comment]
How big is `tif_idx`? Is it 16?

**Comment by william-silversmith** [comment]
Does this graph cover only the upload phase? If there's a download section, that would be sped up by parallel regardless of alignment status.

**Comment by falkben** [comment]
That graph is creating a (small) precomputed volume on disk.

Here's a plot specific to doing the upload phase to `S3`

![image](https://user-images.githubusercontent.com/653031/49970076-3f2c1c00-fef8-11e8-86db-80626c47f410.png)

In both cases, there's a speed up with parallel=True or parallel=(some value >1)

Haven't tested download yet.

**Comment by falkben** [comment]
It seems to use `parallel` for the chunk aligned portion just fine (remember, I'm chunking at every 16 slices and uploading them all at that time).

**Comment by william-silversmith** [comment]
I'm a little confused, I thought that the way your upload is offset meant that there is no aligned core since a single 16 slice operation spans two chunks in the Z dimension. Where is there aligned data?

**Comment by william-silversmith** [comment]
I managed to reproduce the issue and resolved it in #168. I'm going to release it in cloud-volume==0.40.3 We can still discuss performance on the outer shell, but maybe it belongs in another issue.
